glommio/src/read_write_lock.rs

```diff
@@ -0,0 +1,1214 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the
+// MIT/Apache-2.0 License, at your convenience
+
+//! Read-write locks.
+//!
+//! Provides functionality similar to the ['std::sync::RwLock'] except that lock can not be poisoned.
+//!
+//! # Examples
+//!
+//! ```
+//! use glommio::read_write_lock::ReadWriteLock;
+//! use glommio::LocalExecutor;
+//! let lock = ReadWriteLock::new(5);
+//! let ex = LocalExecutor::make_default();
+//!
+//! ex.run( async move {
+//!    // many reader locks can be held at once
+//!    {
+//!        let r1 = lock.read().await.unwrap();
+//!        let r2 = lock.read().await.unwrap();
+//!        assert_eq!(*r1, 5);
+//!        assert_eq!(*r2, 5);
+//!    } // read locks are dropped at this point
+//!
+//!    // only one write lock may be held, however
+//!    {
+//!        let mut w = lock.write().await.unwrap();
+//!        *w += 1;
+//!        assert_eq!(*w, 6);
+//!    } // write lock is dropped here
+//! });
+//! ```
+//!
+use alloc::rc::Rc;
+use core::fmt::Debug;
+use core::marker::PhantomData;
+use std::cell::{RefCell, UnsafeCell};
+use std::collections::hash_map::Entry::{Occupied, Vacant};
+use std::collections::{HashMap, VecDeque};
+use std::error::Error;
+use std::fmt::{Display, Formatter};
+use std::future::Future;
+use std::ops::{Deref, DerefMut};
+use std::pin::Pin;
+use std::task::{Context, Poll, Waker};
+
+#[derive(Clone, Copy, Eq, PartialEq, Hash, Debug)]
+struct WaiterId(u64);
+
+/// A type alias for the result of a lock method which can be blocked.
+pub type ReadWriteLockResult<T> = Result<T, LockClosedError>;
+
+/// A type alias for the result of a nonblocking locking method.
+pub type TryReadWriteLockResult<T> = Result<T, TryReadWriteLockError>;
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+enum WaiterKind {
+    READER,
+    WRITER,
+}
+
+///Error which indicates that ReadWriteLock is closed and can not be use
+///to request lock access
+pub struct LockClosedError;
+
+impl Error for LockClosedError {}
+
+impl Display for LockClosedError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        f.write_str("Lock already closed")
```

Comment:
Here and elsewhere

```suggestion
        f.write_str("lock already closed")
```

https://github.com/rust-lang/api-guidelines/commit/201d9036dbe8640eef6054d2d9632ea0a0532ab6

If you are wondering why `io::Error` looks like `"Permission denied"`, that's because they use C library function to covert errno to message, and that is uppercase for Linux (but lowercase for windwos).